Marlin/src/HAL/HAL_STM32/persistent_store_sdcard.cpp

```diff
@@ -0,0 +1,104 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2019 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (c) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/**
+ * HAL for stm32duino.com based on Libmaple and compatible (STM32F1)
+ * Implementation of EEPROM settings in SD Card
+ */
+
+#ifdef TARGET_STM32F4
+
+#include "../../inc/MarlinConfig.h"
+
+#if ENABLED(EEPROM_SETTINGS) && NONE(FLASH_EEPROM_EMULATION, SPI_EEPROM, I2C_EEPROM)
+
+#include "../shared/persistent_store_api.h"
+
+#ifndef E2END
+  #define E2END 0xFFF // 4KB
+#endif
+#define HAL_EEPROM_SIZE (E2END + 1) // 16KB
+
+#define _ALIGN(x) __attribute__ ((aligned(x))) // SDIO uint32_t* compat.
+static char _ALIGN(4) HAL_eeprom_data[HAL_EEPROM_SIZE];
+
+#if ENABLED(SDSUPPORT)
+
+  #include "../../sd/cardreader.h"
+
+  #define EEPROM_FILENAME "eeprom.dat"
+
+  bool PersistentStore::access_start() {
+    if (!card.isDetected()) return false;
+
+    SdFile file, root = card.getroot();
+    if (!file.open(&root, EEPROM_FILENAME, O_RDONLY))
+      return false;
+
+    int16_t bytes_read = file.read(HAL_eeprom_data, HAL_STM32F4_EEPROM_SIZE);
```

Comment:
you left some HAL_STM32F4_EEPROM_SIZE

Reply:
#define HAL_STM32F4_EEPROM_SIZE (E2END + 1) // 16KB

#define _ALIGN(x) __attribute__ ((aligned(x))) // SDIO uint32_t* compat.
static char _ALIGN(4) HAL_eeprom_data[HAL_STM32F4_EEPROM_SIZE];
@tpruvot You should fix this

Reply:
will try, but ive no boards using "STM32" hal

Reply:
done, in PR #14844 ... the alignment is for the SDIO transferts, but unsure the STM32 has ported it...

Reply:
It seems that in the latest firmware, stm32f103xxx chip board, using SD to simulate EEPROM seems to have a bug, that is, it needs to create a new EEPROM.DAT file in SD card first.But there is nothing wrong with the previous download. I am not sure what changes you have made

Reply:
see #14845